# HG changeset patch
# User Michael Goffioul <michael.goffioul@gmail.com>
# Date 1344193493 -3600
#      Sun Aug 05 20:04:53 2012 +0100
# Branch gui
# Node ID 3d7a7ae53bbfda893f72a0a0312e96f0bf14fa03
# Parent  7279c6fc3acb84d4df35efbf877378c7da92249b
Further integration of GUI in the build system.

* Makefile.am (SUBDIRS): Add gui conditionally.
* configure.ac (AC_CHECK_PROGS(rcc)): Check for new program.
(AC_ARG_ENABLE(gui): New configure flag.
* gui/src/Makefile.am: Add support for resource file compilation.
Use nodist_XXX_SOURCES for MOC files.

diff --git a/Makefile.am b/Makefile.am
--- a/Makefile.am
+++ b/Makefile.am
@@ -65,16 +65,20 @@ EXTRA_DIST = \
   etc/gdbinit \
   run-octave.in
 
 include m4/module.mk
 
 # Subdirectories in which to run `make all'.
 SUBDIRS = libgnu libcruft liboctave src scripts @DOCDIR@ examples test
 
+if OCTAVE_GUI
+SUBDIRS += gui
+endif
+
 if ! AMCOND_BUILD_DOCS
 dist-hook:
 	echo "Documentation disabled.  Cannot package distribution!" ; exit 1;
 endif
 
 BUILT_SOURCES = run-octave
 
 noinst_SCRIPTS = run-octave
diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -2129,16 +2129,17 @@ fi
 
 AC_SUBST(WARN_CFLAGS)
 AC_SUBST(WARN_CXXFLAGS)
 
 ### GUI/Qt related tests.
 
 AC_CHECK_PROGS(MOC, [moc moc-qt4 moc-qt5])
 AC_CHECK_PROGS(UIC, [uic uic-qt4 uic-qt5])
+AC_CHECK_PROGS(RCC, [rcc])
 AC_CHECK_PROGS(QMAKE, [qmake qmake-qt4 qmake-qt5])
 
 AC_SUBST(MOC)
 AC_SUBST(UIC)
 
 QT_INCDIR=`$QMAKE -query | sed -n -e 's/^QT_INSTALL_HEADERS://p'`
 QT_LIBDIR=`$QMAKE -query | sed -n -e 's/^QT_INSTALL_LIBS://p'`
 
@@ -2150,16 +2151,24 @@ case "$canonical_host_type" in
   *)
     win32_terminal=no
     AC_CHECK_HEADERS([pty.h libutil.h util.h])
     AC_SEARCH_LIBS([openpty], [util], [AC_DEFINE([HAVE_OPENPTY], [], [Define whether openpty exists])])
     ;;
 esac
 AM_CONDITIONAL([WIN32_TERMINAL], [test x$win32_terminal = xyes])
 
+octave_gui=
+AC_ARG_ENABLE(gui,
+  [AS_HELP_STRING([--enable-gui], [build GUI (default is no)])],
+  [if test "$enableval" = yes; then
+     octave_gui=yes
+   fi], [])
+AM_CONDITIONAL([OCTAVE_GUI], [test x$octave_gui = xyes])
+
 ### Run configure in subdirectories.
 
 export CC
 export CXX
 export F77
 
 ### Some things to add to the bottom of config.h.
 
diff --git a/gui/src/Makefile.am b/gui/src/Makefile.am
--- a/gui/src/Makefile.am
+++ b/gui/src/Makefile.am
@@ -6,16 +6,19 @@ endef
 
 SRCDIRS = m-editor qtinfo octave-adapter \
 	  ../qterminal/libqterminal/unix ../qterminal/libqterminal/win32 ../qterminal/libqterminal ../qterminal
 $(foreach dir,. $(SRCDIRS),$(eval $(call MOC_template,$(dir))))
 
 ui_%.h: %.ui
 	@UIC@ -o $@ $<
 
+qrc_%.cpp: %.qrc
+	@RCC@ -o $@ $<
+
 bin_PROGRAMS = octave-gui
 
 octave_gui_MOC = \
 		 moc_documentation-dockwidget.cpp \
 		 moc_files-dockwidget.cpp \
 		 moc_history-dockwidget.cpp \
 		 moc_main-window.cpp \
 		 moc_octave-qt-event-listener.cpp \
@@ -32,16 +35,18 @@ octave_gui_MOC = \
 		 octave-adapter/moc_octave-main-thread.cpp \
 		 qtinfo/moc_parser.cpp \
 		 qtinfo/moc_webinfo.cpp
 octave_gui_UI = \
 		settings-dialog.ui \
 		welcome-wizard.ui
 octave_gui_UI_H = $(patsubst %.ui,ui_%.h,$(octave_gui_UI))
 
+octave_gui_RC = qrc_resource.cpp
+
 octave_gui_SOURCES = \
 		     documentation-dockwidget.cc \
 		     files-dockwidget.cc \
 		     history-dockwidget.cc \
 		     main-window.cc \
 		     octave-gui.cc \
 		     octave-qt-event-listener.cc \
 		     resource-manager.cc \
@@ -52,18 +57,18 @@ octave_gui_SOURCES = \
 		     workspace-view.cc \
 		     m-editor/file-editor.cc \
 		     m-editor/file-editor-tab.cc \
 		     m-editor/find-dialog.cc \
 		     m-editor/lexer-octave-gui.cc \
 		     octave-adapter/octave-link.cc \
 		     octave-adapter/octave-main-thread.cc \
 		     qtinfo/parser.cc \
-		     qtinfo/webinfo.cc \
-		     $(octave_gui_MOC)
+		     qtinfo/webinfo.cc
+nodist_octave_gui_SOURCES = $(octave_gui_MOC) $(octave_gui_RC)
 
 octave_gui_CPPFLAGS = -I@QT_INCDIR@/QtCore \
 		      -I@QT_INCDIR@/QtGui \
 		      -I@QT_INCDIR@/QtNetwork \
 		      -I$(srcdir)/../qterminal/libqterminal \
 		      -I$(srcdir)/m-editor \
 		      -I$(srcdir)/octave-adapter \
 		      -I$(srcdir)/qtinfo \
@@ -103,26 +108,26 @@ noinst_HEADERS = \
 		 octave-adapter/octave-event.h \
 		 octave-adapter/octave-event-observer.h \
 		 octave-adapter/octave-event-listener.h \
 		 octave-adapter/octave-link.h \
 		 octave-adapter/octave-main-thread.h \
 		 qtinfo/parser.h \
 		 qtinfo/webinfo.h
 
-CLEANFILES = $(octave_gui_MOC) $(octave_gui_UI_H)
+CLEANFILES = $(octave_gui_MOC) $(octave_gui_UI_H) $(octave_gui_RC)
 
 clean-local:
 	-if test "$(srcdir)" != "."; then \
 		for d in $(SRCDIRS); do test -d $$d && rmdir $$d; done \
 	 fi
 
 BUILT_SOURCES = $(octave_gui_UI_H)
 
-EXTRA_DIST = $(octave_gui_UI)
+EXTRA_DIST = $(octave_gui_UI) resource.qrc
 
 noinst_LTLIBRARIES = libqterminal.la
 
 noinst_HEADERS += ../qterminal/libqterminal/QTerminal.h \
 		  ../qterminal/libqterminal/QTerminalInterface.h \
 		  ../qterminal/libqterminal/win32/QTerminalColors.h \
 		  ../qterminal/libqterminal/win32/QWinTerminalImpl.h \
 		  ../qterminal/libqterminal/unix/BlockArray.h \
@@ -150,40 +155,40 @@ libqterminal_la_CPPFLAGS = $(DEFS)  \
 			   -I@QT_INCDIR@/QtCore \
 			   -I@QT_INCDIR@/QtGui \
 			   -I$(srcdir)/../qterminal/libqterminal
 
 libqterminal_la_MOC = \
 		      ../qterminal/libqterminal/moc_QTerminal.cpp \
 		      ../qterminal/libqterminal/moc_QTerminalInterface.cpp
 
-libqterminal_la_SOURCES = $(libqterminal_la_MOC)
+nodist_libqterminal_la_SOURCES = $(libqterminal_la_MOC)
 
 if WIN32_TERMINAL
-libqterminal_la_SOURCES += \
-			   ../qterminal/libqterminal/win32/QTerminalColors.cpp \
-			   ../qterminal/libqterminal/win32/QTerminalImpl.cpp
+libqterminal_la_SOURCES = \
+			  ../qterminal/libqterminal/win32/QTerminalColors.cpp \
+			  ../qterminal/libqterminal/win32/QWinTerminalImpl.cpp
 libqterminal_la_MOC += ../qterminal/libqterminal/win32/moc_QWinTerminalImpl.cpp
 else
-libqterminal_la_SOURCES += \
-			   ../qterminal/libqterminal/unix/BlockArray.cpp \
-			   ../qterminal/libqterminal/unix/Emulation.cpp \
-			   ../qterminal/libqterminal/unix/Filter.cpp \
-			   ../qterminal/libqterminal/unix/History.cpp \
-			   ../qterminal/libqterminal/unix/KeyboardTranslator.cpp \
-			   ../qterminal/libqterminal/unix/konsole_wcwidth.cpp \
-			   ../qterminal/libqterminal/unix/kpty.cpp \
-			   ../qterminal/libqterminal/unix/QUnixTerminalImpl.cpp \
-			   ../qterminal/libqterminal/unix/Screen.cpp \
-			   ../qterminal/libqterminal/unix/ScreenWindow.cpp \
-			   ../qterminal/libqterminal/unix/TerminalCharacterDecoder.cpp \
-			   ../qterminal/libqterminal/unix/Vt102Emulation.cpp \
-			   ../qterminal/libqterminal/unix/SelfListener.cpp \
-			   ../qterminal/libqterminal/unix/TerminalModel.cpp \
-			   ../qterminal/libqterminal/unix/TerminalView.cpp
+libqterminal_la_SOURCES = \
+			  ../qterminal/libqterminal/unix/BlockArray.cpp \
+			  ../qterminal/libqterminal/unix/Emulation.cpp \
+			  ../qterminal/libqterminal/unix/Filter.cpp \
+			  ../qterminal/libqterminal/unix/History.cpp \
+			  ../qterminal/libqterminal/unix/KeyboardTranslator.cpp \
+			  ../qterminal/libqterminal/unix/konsole_wcwidth.cpp \
+			  ../qterminal/libqterminal/unix/kpty.cpp \
+			  ../qterminal/libqterminal/unix/QUnixTerminalImpl.cpp \
+			  ../qterminal/libqterminal/unix/Screen.cpp \
+			  ../qterminal/libqterminal/unix/ScreenWindow.cpp \
+			  ../qterminal/libqterminal/unix/TerminalCharacterDecoder.cpp \
+			  ../qterminal/libqterminal/unix/Vt102Emulation.cpp \
+			  ../qterminal/libqterminal/unix/SelfListener.cpp \
+			  ../qterminal/libqterminal/unix/TerminalModel.cpp \
+			  ../qterminal/libqterminal/unix/TerminalView.cpp
 libqterminal_la_MOC += \
 		       ../qterminal/libqterminal/unix/moc_Emulation.cpp \
 		       ../qterminal/libqterminal/unix/moc_Filter.cpp \
 		       ../qterminal/libqterminal/unix/moc_QUnixTerminalImpl.cpp \
 		       ../qterminal/libqterminal/unix/moc_ScreenWindow.cpp \
 		       ../qterminal/libqterminal/unix/moc_SelfListener.cpp \
 		       ../qterminal/libqterminal/unix/moc_TerminalModel.cpp \
 		       ../qterminal/libqterminal/unix/moc_TerminalView.cpp \
